### ISSUE TITLE
Rule F821 false positive in class-scope comprehension

### ISSUE THREAD

**Issue by CarrotManMatt** [astral-sh/ruff#23364]
### Summary

In the following code[^1], ruff reports an `undefined-name` diagnostic on the reference to class `Foo`, during the assignment to `QUX`, despite `Foo` being accessible. `Foo` _is_ well-defined here due to being executed inside the comprehension scope.

All iterators, beyond the first one, in a comprehension are executed inside the comprehension scope, not the class scope. This comprehension scope contains a well-defined reference to the class due to it only being executed _after_ the class has finished being defined, so the name is not undefined. (See https://stackoverflow.com/a/13913933 for a full explanation)

```python
class Foo:
    BAR = [1, 2, 3]
    BAZ = [4, 5, 6]
    QUX = [first, second for first in BAR for second in Foo.BAZ]
```

[^1]: I know that there are other solutions to create the same value for `QUX` (E.g. using `itertools.product()`), but this is just a small example that has the incorrect diagnostic

### Version

ruff 0.15.1

**Comment by ntBre** [comment]
I might be missing something subtle here, but based on running this code in the Python REPL, `Foo` is indeed not defined here:

```pycon
>>> class Foo:
...     BAR = [1, 2, 3]
...     BAZ = [4, 5, 6]
...     QUX = [(first, second) for first in BAR for second in Foo.BAZ]
...
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    class Foo:
    ...<2 lines>...
        QUX = [(first, second) for first in BAR for second in Foo.BAZ]
  File "<python-input-1>", line 4, in Foo
    QUX = [(first, second) for first in BAR for second in Foo.BAZ]
                                                          ^^^
NameError: name 'Foo' is not defined
```

**Comment by CarrotManMatt** [comment]
Apologies, my example code here was simplified a bit too much. The place where I encountered this in the real world used a generator comprehension, not a list comprehension. This seems to affect the undefined-ness of the reference to `Foo` as the generator is only consumed after `Foo` has finished being declared. Sorry I'm not 100% sure on the exact reasons for Python's behaviour here anymore then.

It seems the diagnostic is only a false positive if the comprehension is a generator comprehension as it is not executed until later. Whereas list & dictionary comprehensions will be executed eagerly when `Foo` is correctly not defined yet.

```python
class Foo:
    BAR = [1, 2, 3]
    BAZ = [4, 5, 6]
    QUX = ((first, second) for first in BAR for second in Foo.BAZ)

print(list(Foo.QUX))
```

**Comment by amyreese** [comment]
While I agree this could technically be considered a false-positive for F821, I'd argue there should be another rule to avoid assigning a generator expression to a class var, simply because it can be constructed and iterated exactly once, and any further access will just see an exhausted generator. 😅

**Comment by ntBre** [comment]
Thanks for the updated example! Yeah, I guess I'd consider this a bug in F821 and really in our handling of generator scopes, as there doesn't seem to be any generator-specific code in the rule. We don't emit a diagnostic for a somewhat similar case for a lambda:

```py
class Foo:
    BAZ = [4, 5, 6]
    QUX = lambda: Foo.BAZ
```

Although there's kind of a false negative if you invoke the lambda immediately:

```pycon
>>> class Foo:
...     BAZ = 1
...     x = (lambda: Foo.BAZ)()
...
# ...
NameError: name 'Foo' is not defined
```

https://play.ruff.rs/a5e3236e-968d-4d43-9739-f4b45bece823

But we usually err on the side of false negatives if we have to pick one.

**Comment by CarrotManMatt** [comment]
> avoid assigning a generator expression to a class var

I 100% agree. This was just a minimal example. The real-world code declares the generator as an argument to a method's decorator. This is in the same scope as assigning a class var, but is more useful code

<details>
<summary>
Real-world code usage that results in a false-positive F821 diagnostic.
</summary>

Coincidentally this issue arose in the unit tests for a project I am creating to parse Python pragma comments into an AST (E.g. `# noqa: E501` or `type: ignore[override]`).

```python
class TestRuleCodeNode:
    VALID_VALUES: Final[Sequence[LiteralString]] = (
        "RUF001",
        "override",
        "attr-defined",
        "E501",
    )
    TRAILING_COMMA_NODES: Final[Sequence[ast_nodes.Comma | None]] = (
        *(
            ast_nodes.Comma(leading_whitespace=whitespace_node)
            for whitespace_node in TestCommaNode.LEADING_WHITESPACE_NODES
        ),
        None,
    )
    LEADING_WHITESPACE_NODES: Final[Sequence[ast_nodes.Whitespace | None]] = (
        *(ast_nodes.Whitespace(value) for value in TestWhitespaceNode.VALID_VALUES),
        None,
    )

    @pytest.mark.parametrize(
        ("rule_code_node", "expected_str"),
        (
            (
                ast_nodes.RuleCode(
                    value, trailing_comma, leading_whitespace=leading_whitespace
                ),
                (
                    f"{leading_whitespace if leading_whitespace is not None else ''}{value}{
                        trailing_comma if trailing_comma is not None else ''
                    }"
                ),
            )
            for value in VALID_VALUES  # NOTE: No class reference needed because first iterable in a comprehension is executed in the class scope.
            for trailing_comma in TestRuleCodeNode.TRAILING_COMMA_NODES  # noqa: F821
            for leading_whitespace in TestRuleCodeNode.LEADING_WHITESPACE_NODES  # noqa: F821
        ),
    )
    def test_str(
        self, rule_code_node: ast_nodes.RuleCode, expected_str: LiteralString
    ) -> None:
        """"""
        assert str(rule_code_node) == expected_str
```
</details>

**Comment by corevibe555** [comment]
@ntBre Looks like the issue is still open.
Can I work on this?
Thanks

**Comment by corevibe555** [comment]
OK, I think the issue still exists.
I can create PR if you are OK.
